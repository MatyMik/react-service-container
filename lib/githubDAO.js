let instance;
export default class GithubDAO {
  static getInstance() {
    if (!instance) {
      instance = new GithubDAO();
    }
    return instance;
  }

  async listRepos() {
    const res = await fetch("https://api.github.com/repositories");
    const data = await res.json();
    return data.map((repo) => ({
      id: repo.id,
      name: repo.full_name,
      link: repo.html_url,
    }));
  }
}
