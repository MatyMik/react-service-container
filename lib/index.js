import React, { useContext } from "react";

export class ServiceContainerEnvironment {
  constructor(parent = null) {
    this.parent = parent;
    this.providers = new Map();
  }

  get(serviceToken) {
    if (this.providers.has(serviceToken)) {
      const initFn = this.providers.get(serviceToken);
      return initFn();
    }

    if (this.parent != null) {
      // This will recursively call each parent container, until the base environment is hit in
      // which case the base case is thrown below.
      return this.parent.get(serviceToken);
    }

    throw new Error(`Missing provider for service ${serviceToken}`);
  }

  add(provider) {
    if (!("provide" in provider)) {
      throw new Error('Missing "provide" key');
    }

    let initFn;
    switch (true) {
      case "useValue" in provider:
        const value = provider.useValue;
        initFn = () => value;
        break;
      case "useClass" in provider:
        const Ctor = provider.useClass;
        initFn = () => new Ctor();
        break;
      case "useFactory" in provider:
        initFn = provider.useFactory;
        break;
      case "useExisting" in provider:
        if (!this.providers.has(provider.useExisting)) {
          throw new Error("(TODO: Good message about non-existent existing)");
        }
        initFn = this.providers.get(provider.useExisting);
        break;
      default:
        throw new Error("(TODO: Good errors)");
    }

    this.providers.set(provider.provide, initFn);
  }
}

function readonlyProxy(env) {
  const proxy = {
    get(...args) {
      return env.get(...args);
    },
  };
  return proxy;
}

export const ServiceContainerContext = React.createContext(
  readonlyProxy(new ServiceContainerEnvironment())
);

export function ServiceContainer({ providers, children }) {
  const parent = useContext(ServiceContainerContext);
  const container = buildContainer(providers, parent);

  return (
    <ServiceContainerContext.Provider value={readonlyProxy(container)}>
      {children}
    </ServiceContainerContext.Provider>
  );
}

export function useService(serviceToken) {
  const container = useContext(ServiceContainerContext);
  return container.get(serviceToken);
}

function buildContainer(providers, parent) {
  const container = new ServiceContainerEnvironment(parent);
  addProviders(container, providers);
  return container;
}

function addProviders(container, providers) {
  normalize(providers).forEach((provider) => {
    container.add(provider);
  });
}

function normalize(providers) {
  return providers.map((provider) => {
    const assumeClassShorthand = !("provide" in provider);
    if (assumeClassShorthand) {
      return { provide: provider, useClass: provider };
    }
    return provider;
  });
}
