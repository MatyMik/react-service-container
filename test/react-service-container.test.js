import { ServiceContainer, useService, ServiceContainerContext } from "../lib";
import React from "react";
import renderer from "react-test-renderer";

class Dep {
  fn() {}
}

const Component = () => {
  const dep = useService(Dep);
  return <p>{dep.fn()}</p>;
};

let mock;
beforeEach(() => {
  mock = jest.fn();
});

test("useValue", () => {
  renderWithProviders([{ provide: Dep, useValue: { fn: mock } }]);
  expect(mock).toHaveBeenCalled();
});

test("useClass", () => {
  class MockClass {
    fn() {
      mock();
    }
  }

  renderWithProviders([{ provide: Dep, useClass: MockClass }]);
  expect(mock).toHaveBeenCalled();
});

test("useFactory", () => {
  const factory = () => ({
    fn: mock,
  });

  renderWithProviders([{ provide: Dep, useFactory: factory }]);
  expect(mock).toHaveBeenCalled();
});

test("useExisting", () => {
  class Existing {
    fn() {
      mock();
    }
  }

  renderWithProviders([
    { provide: Existing, useClass: Existing },
    { provide: Dep, useExisting: Existing },
  ]);
  expect(mock).toHaveBeenCalled();
});

test("Class shorthand", () => {
  class MockDep {
    fn() {
      mock();
    }
  }

  const Component = () => {
    const mockDep = useService(MockDep);
    return <p>{mockDep.fn()}</p>;
  };

  renderer.create(
    <ServiceContainer providers={[MockDep]}>
      <Component />
    </ServiceContainer>
  );
  expect(mock).toHaveBeenCalled();
});

test("Injection Tokens", () => {
  const CONFIG = Symbol.for("config");
  const config = { message: "Hello, world!" };

  const Component = () => {
    const { message } = useService(CONFIG);
    return <p>{message}</p>;
  };
  const component = renderer.create(
    <ServiceContainer providers={[{ provide: CONFIG, useValue: config }]}>
      <Component />
    </ServiceContainer>
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test("useExisting with injection tokens", () => {
  const V1 = Symbol.for("v1");
  const VALIAS = Symbol.for("vAlias");
  const v1 = "value 1";

  const Component = () => {
    const v1 = useService(VALIAS);
    return <p>{v1}</p>;
  };

  const component = renderer.create(
    <ServiceContainer
      providers={[
        { provide: V1, useValue: v1 },
        { provide: VALIAS, useExisting: V1 },
      ]}
    >
      <Component />
    </ServiceContainer>
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test("Hierarchal injection", () => {
  const parentMock = jest.fn();
  const childMock = jest.fn();
  const overrideMock = jest.fn();
  const mockToBeOverriden = jest.fn();

  class ParentDep {
    fn() {
      parentMock();
    }
  }

  class Child {
    fn() {
      childMock();
    }
  }

  class DepToOverride {
    fn() {
      mockToBeOverriden();
    }
  }

  class Override {
    fn() {
      overrideMock();
    }
  }

  const Component = () => {
    const parent = useService(ParentDep);
    const child = useService(Child);
    const override = useService(DepToOverride);

    return (
      <p>
        {parent.fn()}
        {child.fn()}
        {override.fn()}
      </p>
    );
  };

  renderer.create(
    <ServiceContainer providers={[ParentDep, DepToOverride]}>
      <ServiceContainer
        providers={[Child, { provide: DepToOverride, useClass: Override }]}
      >
        <Component />
      </ServiceContainer>
    </ServiceContainer>
  );

  expect(parentMock).toHaveBeenCalled();
  expect(childMock).toHaveBeenCalled();
  expect(overrideMock).toHaveBeenCalled();
  expect(mockToBeOverriden).not.toHaveBeenCalled();
});

test("Class components", () => {
  let mock = jest.fn();

  class Dep {
    fn() {
      mock();
    }
  }

  class Component extends React.Component {
    render() {
      const dep = this.context.get(Dep);
      return <p>{dep.fn()}</p>;
    }
  }
  Component.contextType = ServiceContainerContext;

  renderer.create(
    <ServiceContainer providers={[Dep]}>
      <Component />
    </ServiceContainer>
  );
  expect(mock).toHaveBeenCalled();
});

test.todo("Descriptive error message when useService cannot find context");

test.todo("Descriptive error when missing provider");

test.todo("Descriptive error when provider function throws");

test.todo("Descriptive error message when useExisting lookup fails");

test.todo("Warn when overriding existing provider (in development?)");

function renderWithProviders(providers) {
  return renderer.create(
    <ServiceContainer providers={providers}>
      <Component />
    </ServiceContainer>
  );
}
