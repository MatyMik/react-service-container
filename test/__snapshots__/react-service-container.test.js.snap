// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Injection Tokens 1`] = `
<p>
  Hello, world!
</p>
`;

exports[`useExisting with injection tokens 1`] = `
<p>
  value 1
</p>
`;
